helix-view/src/document.rs

```diff
@@ -112,11 +129,59 @@ pub fn canonicalize_path(path: &Path) -> std::io::Result<PathBuf> {
     std::env::current_dir().map(|current_dir| normalize_path(&current_dir.join(path)))
 }
 
+pub fn auto_detect_line_ending(doc: &Rope) -> LineEnding {
+    // based on https://github.com/cessen/led/blob/27572c8838a1c664ee378a19358604063881cc1d/src/editor/mod.rs#L88-L162
+
+    let mut ending = LineEnding::None;
+    for line in doc.lines().take(1) { // check first line only - unsure how sound this is
+        ending = match line.len_chars() {
+            1 => { let g = RopeGraphemes::new(line.slice((line.len_chars() - 1)..))
+                .last()
+                .unwrap();
+            rope_slice_to_line_ending(&g)}
+            n if n > 1 => { let g = RopeGraphemes::new(line.slice((line.len_chars() - 2)..))
+                .last()
+                .unwrap();
+            rope_slice_to_line_ending(&g) }
+            _ => LineEnding::None
+
+        }
+    }
+    ending
+}
+
+pub fn rope_slice_to_line_ending(g: &RopeSlice) -> LineEnding {
+    if let Some(text) = g.as_str() {
+        str_to_line_ending(text)
+    } else if g == "\u{000D}\u{000A}" {
+        LineEnding::Crlf
+    } else {
+        // Not a line ending
+        LineEnding::None
+    }
+}
+
+pub fn str_to_line_ending(g: &str) -> LineEnding {
+    match g {
+        "\u{000D}\u{000A}" => LineEnding::Crlf,
+        "\u{000A}" => LineEnding::LF,
+        "\u{000B}" => LineEnding::VT,
+        "\u{000C}" => LineEnding::FF,
+        "\u{000D}" => LineEnding::CR,
+        "\u{0085}" => LineEnding::Nel,
+        "\u{2028}" => LineEnding::LS,
+        "\u{2029}" => LineEnding::PS,
+
+        // Not a line ending
+        _ => LineEnding::None,
+    }
+}
+
 use helix_lsp::lsp;
 use url::Url;
 
 impl Document {
-    pub fn new(text: Rope) -> Self {
+    pub fn new(text: Rope, _line_ending: LineEnding) -> Self {
```

Comment:
Remove `line_ending` from `new`, make `Document::new` default to the system default (declare a  `const DEFAULT_LINE_ENDING` that uses `cfg` attrs to be set to `\r\n` if `target_os` is `windows`, else `\n`).

`Document::load` should detect the line ending, then change `doc.line_ending`.